### PR TITLE
LocalScanner: De-duplicate scan results retrieved from the storage

### PR DESCRIPTION
**Pull request by fviernau**
The PostgresStorage handling is currently implemented such that multiple
scan results may get stored for a single identifier, scanner details
provenance tuple. This situation is in fact not unlikely as the time
interval between (1) the look-up of a stored scan result for a package
and (2) the actual scan executed due to the failed look-up may be
relatively large. I've observed this interval to span more than three
hours and the scan storage I'm using for some packages has even six
equal scan results which all do end-up in the OrtResults, increasing its
size unnecessarily.

While the above mentioned time interval should be reduced, and ideally
it also should be made impossible that multiple scan results with equal
identifier, provenance and scanner details can be stored in the first
place, this change only ensures that such duplicates are not put into
OrtResults. The fix could alternatively be implemented in the
PostgresStorage class by modifying the SQL query - I've decided against
this as that would be storage backend specific and would either lead to
a quite complex and initially slow query or require a larger re-thinking
of the database schema.

Signed-off-by: Frank Viernau <frank.viernau@here.com>

### CODE REVIEW

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 1]
```diff
@@ -69,6 +69,8 @@ import kotlinx.coroutines.asCoroutineDispatcher
 import kotlinx.coroutines.async
 import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.withContext
+import org.ossreviewtoolkit.model.RemoteArtifact
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:72]: Nit: Blank line before this one.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 2]
```diff
@@ -455,3 +457,38 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
...
+  */
+private fun ScanResultContainer.deduplicateScanResults(): ScanResultContainer {
+    // Use vcsInfo and sourceArtifact instead of provenance in order to ingore the download time and original VCS info.
+    data class Key(
+        val id: Identifier,
+        val vcsInfo: VcsInfo?,
+        val sourceArtifact: RemoteArtifact?,
+        val scannerDetails: ScannerDetails
+    )
+    fun ScanResult.key() = Key(id, provenance.vcsInfo, provenance.sourceArtifact, scanner)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:484]: Nit: Blank line before this one.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 3]
```diff
@@ -455,3 +457,38 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
...
+ * [OrtResult] produced by this scanner.
+ *
+ * The time interval between a failing look-up of the cache entry and the resulting scan with the following store
+ * operation can be relatively large. Thus this [LocalScanner] is prone to adding duplicate scan results if scans
+ * are run in parallel. In particular the [PostgresStorage] allows adding duplicate tuples
+ * (identifier, provenance, scanner details) which probably should be made unique.
+ *
+ * TODO:
+ *
+ * 1. Minimize the time between the failing look-up and the curresponding store operation mentioned.
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:472]: "corresponding"

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 4]
```diff
@@ -455,3 +457,38 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
...
+ * 2. Make the tuples (identifier, provenance, scanner details) unique (dis-regarding provenance.downloadTime), at
+ * least in [PostgresStroage].
+  */
+private fun ScanResultContainer.deduplicateScanResults(): ScanResultContainer {
+    // Use vcsInfo and sourceArtifact instead of provenance in order to ingore the download time and original VCS info.
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:477]: "ignore"

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 5]
```diff
@@ -455,3 +457,38 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
...
+    data class Key(
+        val id: Identifier,
+        val vcsInfo: VcsInfo?,
+        val sourceArtifact: RemoteArtifact?,
+        val scannerDetails: ScannerDetails
+    )
+    fun ScanResult.key() = Key(id, provenance.vcsInfo, provenance.sourceArtifact, scanner)
+
+    val deduplicatedResults = results.associateBy { it.key() }.values.toList()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:486]: How about `results.distinctBy { it.key() }`?

File: scanner/src/main/kotlin/LocalScanner.kt (hunk 1 above)
Comment (fviernau) [reply scanner/src/main/kotlin/LocalScanner.kt:72]: Done.

File: scanner/src/main/kotlin/LocalScanner.kt (hunk 2 above)
Comment (fviernau) [reply scanner/src/main/kotlin/LocalScanner.kt:484]: Done.

File: scanner/src/main/kotlin/LocalScanner.kt (hunk 3 above)
Comment (fviernau) [reply scanner/src/main/kotlin/LocalScanner.kt:472]: Done.

File: scanner/src/main/kotlin/LocalScanner.kt (hunk 4 above)
Comment (fviernau) [reply scanner/src/main/kotlin/LocalScanner.kt:477]: Done.

File: scanner/src/main/kotlin/LocalScanner.kt (hunk 5 above)
Comment (fviernau) [reply scanner/src/main/kotlin/LocalScanner.kt:486]: Done.

**Pull request merged**